pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -148,56 +148,166 @@ func NewFlowExporter(connStore *connections.ConntrackConnectionStore, records *f
+	return exp.idleFlowTimeout
+}
+
+func (exp *flowExporter) sendFlowRecords() error {
+	exp.expirePriorityQueue.Mutex.Lock()
+	defer exp.expirePriorityQueue.Mutex.Unlock()
+
+	if exp.expirePriorityQueue.Len() == 0 {
+		return nil
+	}
+	currTime := time.Now()
+	for exp.expirePriorityQueue.Len() > 0 {
+		topItem := exp.expirePriorityQueue.Peek()
+		// If the top item is not timeout, then we do not have to check the following items.
+		if topItem.ActiveExpireTime.After(currTime) && topItem.IdleExpireTime.After(currTime) {
+			break
+		}
+		pqItem := heap.Pop(exp.expirePriorityQueue).(*flowexporter.ItemToExpire)
+		// The pqItem contains a flow record.
+		if !pqItem.IsDenyConn {
+			if pqItem.IsActive {
+				// Send a flow record if it is active and time out.
+				// TODO: more records per data set will be supported when go-ipfix supports size check when adding records
+				if err := exp.addConnToSet(pqItem.Conn, false); err != nil {
+					return err
+				}
+				if _, err := exp.sendDataSet(); err != nil {
+					return err
+				}
+				exp.numDataSetsSent = exp.numDataSetsSent + 1
+			}
+			if flowexporter.IsConnectionDying(&pqItem.Conn) {
+				// If the connection is in dying state or connection is not in conntrack
+				// table, we set the DyingAndDoneExport flag to do the deletion later.
+				pqItem.DyingAndDoneExport = true
+			} else {
+				if pqItem.IdleExpireTime.Before(currTime) && (pqItem.Conn.OriginalPackets <= pqItem.PrevPackets) && (pqItem.Conn.ReversePackets <= pqItem.PrevReversePackets) {
+					pqItem.IsActive = false
+				}
+				exp.ValidateAndUpdateStats(pqItem)
+				if pqItem.IdleExpireTime.Before(currTime) {
+					pqItem.IdleExpireTime = currTime.Add(exp.idleFlowTimeout)
+				}
+				if pqItem.ActiveExpireTime.Before(currTime) {
+					pqItem.ActiveExpireTime = currTime.Add(exp.activeFlowTimeout)
+				}
+				heap.Push(exp.expirePriorityQueue, pqItem)
+			}
+			klog.V(4).InfoS("Record for connection sent successfully", "flowKey", pqItem.ConnKey, "connection", pqItem.Conn)
+		}
+		// The pqItem contains a deny connection.
+		if pqItem.IsDenyConn {
+			if err := exp.addConnToSet(pqItem.Conn, true); err != nil {
+				return err
+			}
+			if _, err := exp.sendDataSet(); err != nil {
+				return err
+			}
+			exp.numDataSetsSent = exp.numDataSetsSent + 1
+			klog.V(4).InfoS("Record for deny connection sent successfully", "flowKey", pqItem.ConnKey, "connection", pqItem.Conn)
+			// Reset the connection's active expire time if it is active time out.
+			if pqItem.Conn.OriginalPackets > pqItem.PrevPackets && pqItem.ActiveExpireTime.Before(currTime) {
+				exp.ValidateAndUpdateStats(pqItem)
+				pqItem.ActiveExpireTime = currTime.Add(exp.activeFlowTimeout)
+				heap.Push(exp.expirePriorityQueue, pqItem)
+			}
+			// If the connection is idle time out, we set the DyingAndDoneExport
+			// flag to do the deletion later.
+			if pqItem.IdleExpireTime.Before(currTime) {
+				pqItem.DyingAndDoneExport = true
```

Comment:
Currently reusing the `DyingAndDoneExport` flag to mark the deny connection need to be deleted. The field naming may be somehow unrelated. Seeking for advice.